internal/controller/controllers/installenv_controller.go

```diff
@@ -0,0 +1,218 @@
+		imageState.State = ImageBeingCreated
+		imageState.StateInfo = adiiov1alpha1.ImageStateCreated
+	}
+
+	return r.updateStatusAndReturnResult(ctx, installEnv, imageInfo, imageState, inventoryErr)
+}
+
+func (r *InstallEnvReconciler) getClusterDeploymentByRef(ctx context.Context, ref *adiiov1alpha1.ClusterReference) (*hivev1.ClusterDeployment, error) {
+	key := types.NamespacedName{
+		Name:      ref.Name,
+		Namespace: ref.Namespace,
+	}
+	clusterDeployment := &hivev1.ClusterDeployment{}
+	if err := r.Get(ctx, key, clusterDeployment); err != nil {
+		return nil, newKubeAPIError(
+			errors.Wrapf(
+				err,
+				fmt.Sprintf(
+					"failed to find clusterDeployment with name %s in namespace %s",
+					ref.Name, ref.Namespace)),
+			k8serrors.IsNotFound(err))
+	}
+	return clusterDeployment, nil
+}
+
+func (r *InstallEnvReconciler) updateStatusAndReturnResult(
+	ctx context.Context,
+	installEnv *adiiov1alpha1.InstallEnv,
+	imageInfo *models.ImageInfo,
+	imageState ImageState,
+	err error) (ctrl.Result, error) {
+
+	var res ctrl.Result
+
+	if imageBeingCreated(err) {
+		imageState.State = ImageCreated
+		imageState.StateInfo = adiiov1alpha1.ImageStateCreated
+		// Clear up the error stateInfo while installEnv is being created
+		err = nil
+		r.Log.Infof("Image %s being prepared for cluster %s stateInfo: %s",
+			installEnv.Name, installEnv.ClusterName, imageState.StateInfo)
+	} else if isClientError(err) {
+		imageState.State = ImageCreationError // TODO before merge: should we do this?
+		imageState.StateInfo += ": " + err.Error()
+	} else if err != nil {
+		imageState.State = ImageCreationError
+		imageState.StateInfo += ": internal error"
+		res.Requeue = true
+	}
+	r.setStateAndStateInfo(adiiov1alpha1.ImageProgressCondition, imageState, installEnv)
+
+	if err == nil && imageInfo != nil {
+		installEnv.Status.ISODownloadURL = imageInfo.DownloadURL
+	} else if err != nil {
+		r.Log.WithError(err).Error("installEnv reconcile failed")
+	}
+
+	if updateErr := r.Status().Update(ctx, installEnv); updateErr != nil {
+		r.Log.WithError(updateErr).Error("failed to update installEnv status")
+		res.Requeue = true
+		return res, nil
+	}
+
+	return res, nil
+}
+
+func (r *InstallEnvReconciler) setStateAndStateInfo(conditionType adiiov1alpha1.InstallEnvConditionType,
+	imageState ImageState, installEnv *adiiov1alpha1.InstallEnv) {
+	r.setCondition(adiiov1alpha1.InstallEnvCondition{
+		Type:               conditionType,
+		Status:             corev1.ConditionUnknown,
+		LastProbeTime:      metav1.Time{Time: time.Now()},
+		LastTransitionTime: metav1.Time{Time: time.Now()},
+		Reason:             imageState.State,
+		Message:            imageState.StateInfo,
+	}, &installEnv.Status.Conditions)
+}
+
+func (r *InstallEnvReconciler) setCondition(condition adiiov1alpha1.InstallEnvCondition, conditions *[]adiiov1alpha1.InstallEnvCondition) {
+	if index := findInstallEnvConditionIndexByReason(condition.Reason, conditions); index >= 0 {
```

Comment:
Shouldn't we be finding conditions by type?

Reply:
Yes. Find in the update I have just pushed.